src/metabase/api/collection.clj

```diff
@@ -79,64 +81,159 @@
+                [:= :collection_id (:id collection)]
+                [:= :archived (boolean archived?)]]})
+
+(defmethod post-process-collection-children :snippet
+  [_ rows]
+  (for [row rows]
+    (dissoc row :description :collection_position :display)))
+
+(defmethod collection-children-query :card
+  [_ collection {:keys [archived? pinned-state]}]
+  (-> {:select [:id :name :description :collection_position :display [(hx/literal "card") :model]]
+       :from   [[Card :c]]
+       :where  [:and
+                [:= :collection_id (:id collection)]
+                [:= :archived (boolean archived?)]]}
+      (h/merge-where (pinned-state->clause pinned-state))))
+
+(defmethod post-process-collection-children :card
+  [_ rows]
+  (let [last-edits (last-edit/fetch-last-edited-info {:card-ids (->> rows (map :id))})]
+    (for [row (hydrate rows :favorite)]
+        (if-let [edit-info (get-in last-edits [:card (:id row)])]
+          (assoc row :last-edit-info edit-info)
+          row))))
+
+(defmethod collection-children-query :dashboard
+  [_ collection {:keys [archived? pinned-state]}]
+  (-> {:select [:id :name :description :collection_position [nil :display] [(hx/literal "dashboard") :model]]
+       :from   [[Dashboard :d]]
+       :where  [:and
+                [:= :collection_id (:id collection)]
+                [:= :archived (boolean archived?)]]}
+      (h/merge-where (pinned-state->clause pinned-state))))
+
+(defmethod post-process-collection-children :dashboard
+  [_ rows]
+  (let [last-edits (last-edit/fetch-last-edited-info {:dashboard-ids (->> rows (map :id))})]
+    (for [row (hydrate rows :favorite)]
+      (let [res (dissoc row :display)]
+        (if-let [edit-info (get-in last-edits [:dashboard (:id res)])]
+          (assoc res :last-edit-info edit-info)
+          res)))))
+
+(defmethod collection-children-query :collection
+  [_ collection {:keys [archived? collection-namespace pinned-state]}]
+  (-> (assoc (collection/effective-children-query
+           collection
+           [:= :archived archived?]
+           [:= :namespace (u/qualified-name collection-namespace)])
+         ; We get from the effective-children-query a normal set of columns selected:
+         ; want to make it fit the others to make UNION ALL work
+         :select [:id
+                  :name
+                  :description
+                  [nil :collection_position]
+                  [nil :display]
+                  [(hx/literal "collection") :model]])
+      ; the nil indicates that collections are never pinned.
+      (h/merge-where (pinned-state->clause pinned-state nil))))
+
+(defmethod post-process-collection-children :collection
+  [_ rows]
+  (for [row rows]
+    ; Go through this rigamarole instead of hydration because we
+    ; don't get models back from ulterior over-query
+    ; Previous examination with logging to DB says that there's no N+1 query for this.
+    ; However, this was only tested on H2 and Postgres
```

Comment:
https://guide.clojure.style/#two-semicolons-for-code-fragment

Emacs/CIDER is going to auto-indent these somewhere crazy if you only use a single semicolon. Use two if they're at the start of the line